### PR TITLE
Numeric Extraction Benchmarking Framework

### PR DESCRIPTION
**Pull request by fmdelgado**
I've initiated this pull request for benchmarking LLMs against numerical data extraction/reasoning tasks to invite early feedback and foster  improvemen. Please review and share your thoughts or modifications, as this is an ongoing process.

Key Developments and Issues:

- Prompt Template & Output Formatting Across LLMs: Currently, all language learning models (LLMs) utilize the same prompt template and output formatting. This uniformity is crucial for comparative analysis, but it also introduces some challenges.

- Output Formatting: I've observed occasional failures in output formatting. Interestingly, after multiple attempts, the system seems to correctly retrieve numbers and place them into a Python dictionary. This inconsistency needs to be addressed for reliable performance, e.g. run 5 times.

- For evaluating the numeric extraction capability, we have a dataset of 650 texts, each with known real answers. 

Areas Needing Assistance:

- Separation of JSON: I propose that we need a clear distinction in our evaluation between the LLMs' ability to format output correctly (in JSON) and their proficiency in accurately retrieving numeric information. This separation will help in pinpointing the specific areas each model excels in or needs improvement.

- Pytest: I am not well-versed with pytest for implementing tests in this context. Your expertise or guidance in this area would be highly beneficial.

- New Branch for Implementation: I plan to initiate a new branch where these aspects can be implemented and tested more effectively. This branch will focus on refining the output formatting process and enhancing the accuracy of numeric extraction.

- Specific Scoring for This Dataset: It’s worth noting that the scoring mechanism we're using is tailored for this particular dataset. While it's effective in this context, we might need to adapt or expand it for broader applications.

### CODE REVIEW

**Comment by slobentanzer** [comment]
changed the merge base to the existing `benchmark` branch, we don't need to create a new branch for that.

**Comment by slobentanzer** [comment]
@nilskre I guess we need to go over this case and see how it compares to the automated/encrypted workflow (after the hackathon)

**Comment by nilskre** [comment]
Yes, would be nice to evaluate, whether we can bring these two developments together.

**Review by nilskre — COMMENTED** [review]
Looks good. Due to a 503 on [http://llm.biocypher.org/](http://llm.biocypher.org/) I was not able to test it practically.  

I think that there is still some work left in making it more generally applicable. At the moment it is tailored towards this one dataset.

But I think coming up with a concept that is generally applicable (and to adapt this case), should probably be its own follow-up task (and PR).

File: benchmark/numericQA.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
+import pandas as pd
+import requests
+import json
+from word2number import w2n
```
[the remark this answers is not in the document]
Comment (nilskre) [reply benchmark/numericQA.py:4]: Added the dependency

File: benchmark/numericQA.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+from xinference.client import Client
+from langchain.llms import Xinference
+from langchain.output_parsers import StructuredOutputParser, ResponseSchema
+from langchain.prompts import PromptTemplate
+from tqdm import tqdm
+from sklearn.metrics import accuracy_score, precision_recall_fscore_support
```
[the remark this answers is not in the document]
Comment (nilskre) [reply benchmark/numericQA.py:10]: Added the dependency

**Comment by slobentanzer** [comment]
Agree with this needing more integration into biochatter to be a part of the ecosystem; as is, this really is just a script tailored to the dataset, without biochatter involvement. Also, I feel that testing the edge cases and single representative examples may be more suitable for integrating into the current pytest framework. Also because running all cases is quite resource-intensive for testing only one particular kind of question. For now I am making this a draft to be addressed/integrated later.

**Comment by slobentanzer** [comment]
Closing this as stale.
